python-avd/pyavd/_cv/workflows/verify_inputs.py

```diff
@@ -0,0 +1,203 @@
+    )
+
+    # Populate list of CVDevice with duplicated serial_number values
+    for current_serial_number, device_iterator_object in devices_grouped_by_serial_number:
+        if len(devices_with_current_serial_number := list(device_iterator_object)) > 1:
+            duplicated_serial_number.append(
+                {"duplicated_serial_number": current_serial_number, "devices_with_duplicated_serial_number": devices_with_current_serial_number}
+            )
+
+    # Populate list of CVDevice with duplicated system_mac_address values
+    for current_system_mac_address, device_iterator_object in devices_grouped_by_system_mac_address:
+        if len(devices_with_current_system_mac_address := list(device_iterator_object)) > 1:
+            if devices_with_unset_serial_number := [device for device in devices_with_current_system_mac_address if device.serial_number is None]:
+                duplicated_system_mac_address_unset_serial_number.append(
+                    {
+                        "duplicated_system_mac_address": current_system_mac_address,
+                        "devices_with_duplicated_system_mac_address": devices_with_unset_serial_number,
+                    }
+                )
+            if devices_with_set_serial_number := [device for device in devices_with_current_system_mac_address if device.serial_number is not None]:
+                duplicated_system_mac_address_set_serial_number.append(
+                    {
+                        "duplicated_system_mac_address": current_system_mac_address,
+                        "devices_with_duplicated_system_mac_address": devices_with_set_serial_number,
+                    }
+                )
+
+    return duplicated_serial_number, duplicated_system_mac_address_unset_serial_number, duplicated_system_mac_address_set_serial_number
+
+
+def identify_duplicated_devices_new(
+    *, devices: list[CVDevice]
+) -> tuple[list[dict[str, str | list[CVDevice]]], list[dict[str, str | list[CVDevice]]], list[dict[str, str | list[CVDevice]]]]:
+    """
+    Process list of CVDevice instances to identify those with overlapping serial_number or system_mac_address.
+
+    Return tuple containing:
+      - List of dictionaries with information about CVDevices with overlapping serial_number.
+      - List of dictionaries with information about CVDevices with overlapping system_mac_address and unset serial_number.
+      - List of dictionaries with information about CVDevices with overlapping system_mac_address and set serial_number.
+    """
+    devices_grouped_by_serial_number = defaultdict(lambda: defaultdict(list))
+    devices_grouped_by_system_mac_address = defaultdict(lambda: defaultdict(list))
+    for device in devices:
+        devices_grouped_by_serial_number[device.serial_number][device.system_mac_address].append(device)
+        devices_grouped_by_system_mac_address[device.system_mac_address][device.serial_number].append(device)
```

Comment:
yeah so you have iterables and these are big lists so less efficient and we cannot do what I wanted here alas (because I missed a point in the logic)